### PR TITLE
[NeMo-UX] Make TE and Apex dependencies optional

### PR DESCRIPTION
**Pull request by ashors1**
# What does this PR do ?

This PR provides a fallback codepath that defaults to pure Pytorch/jit when TE and Apex are not installed. Note that this is specifically designed to work with NeMo 2.0, so there are no guarantees when running with NeMo 1.0 models.

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/nlp/models/language_modeling/megatron_retro_model.py [hunk 1]
```diff
@@ -99,6 +99,9 @@
     TransformerConfig = ApexGuardDefaults
 
     HAVE_MEGATRON_CORE = False
+    from typing import Any
+
+    RetroConfig = Any
```
Comment (dimapihtar) [inline nemo/collections/nlp/models/language_modeling/megatron_retro_model.py:104]: do we want to do the same for all of others mcore's classes we are trying to import under "try" block?
also, why don't use `ApexGuardDefaults`?

File: nemo/lightning/data.py [hunk 2]
```diff
@@ -53,11 +53,16 @@ def setup_microbatch_calculator(
...
-    )
+    try:
+        from apex.transformer.microbatches import ConstantNumMicroBatches
+        from apex.transformer.pipeline_parallel.utils import (
+            _GLOBAL_NUM_MICROBATCHES_CALCULATOR,
+            setup_microbatch_calculator,
+        )
+
+        HAVE_APEX = True
+    except ModuleNotFoundError:
```
Comment (dimapihtar) [inline nemo/lightning/data.py:64]:
```suggestion
    except (ImportError, ModuleNotFoundError):
```

File: nemo/lightning/megatron_parallel.py [hunk 3]
```diff
@@ -112,7 +112,12 @@ def __init__(
         cpu: bool = False,
         convert_module_fn: Optional[Callable[[nn.Module], nn.Module]] = None,
     ) -> None:
-        from apex.transformer.tensor_parallel.layers import set_defaults_if_not_set_tensor_model_parallel_attributes
+        try:
+            from apex.transformer.tensor_parallel.layers import (
+                set_defaults_if_not_set_tensor_model_parallel_attributes,
+            )
+        except ModuleNotFoundError:
```
Comment (dimapihtar) [inline nemo/lightning/megatron_parallel.py:119]:
```suggestion
        except (ImportError, ModuleNotFoundError):
```

File: nemo/lightning/pytorch/plugins/data_sampler.py [hunk 4]
```diff
@@ -61,7 +61,10 @@ def compute_consumed_samples(self, steps_since_resume=0) -> int:
         app_state = AppState()
 
         if self.rampup_batch_size is not None:
-            from apex.transformer.pipeline_parallel.utils import _GLOBAL_NUM_MICROBATCHES_CALCULATOR
+            try:
+                from apex.transformer.pipeline_parallel.utils import _GLOBAL_NUM_MICROBATCHES_CALCULATOR
+            except ModuleNotFoundError:
```
Comment (dimapihtar) [inline nemo/lightning/pytorch/plugins/data_sampler.py:66]:
```suggestion
            except (ImportError, ModuleNotFoundError):
```

File: nemo/collections/llm/gpt/model/base.py [hunk 5]
```diff
@@ -43,12 +43,21 @@ def configure_model(self, tokenizer) -> "MCoreGPTModel":
...
         from megatron.core import parallel_state
-        from megatron.core.models.gpt.gpt_layer_specs import get_gpt_layer_with_transformer_engine_spec
         from megatron.core.models.gpt.gpt_model import GPTModel as MCoreGPTModel
 
+        try:
+            import transformer_engine
+            from megatron.core.models.gpt.gpt_layer_specs import get_gpt_layer_with_transformer_engine_spec
+
+            get_gpt_layer_spec = get_gpt_layer_with_transformer_engine_spec
+        except ImportError:
```
Comment (dimapihtar) [inline nemo/collections/llm/gpt/model/base.py:53]:
```suggestion
        except (ImportError, ModuleNotFoundError):
```

File: nemo/collections/nlp/models/language_modeling/megatron_retro_model.py (hunk 1 above)
Comment (ashors1) [reply nemo/collections/nlp/models/language_modeling/megatron_retro_model.py]: > do we want to do the same for all of others mcore's classes we are trying to import under "try" block?

So this PR is intended to provide a pure pytorch/jit path for NeMo 2.0, so things are not guaranteed to work with NeMo 1.0 models (I have updated the description to reflect this). I added this change because previously, the import of the Retro model in https://github.com/NVIDIA/NeMo/blob/bdb4e89d9ac33d733f8ea7b21552628dda798825/nemo/collections/nlp/models/language_modeling/__init__.py was breaking the 2.0 codepath, but it looks like that's no longer the case anymore. If it makes things clearer, I go ahead and revert this change.

**Review by dimapihtar — APPROVED** [review]

File: examples/llm/megatron_gpt_pretraining.py [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+from pytorch_lightning.loggers import TensorBoardLogger
+
+from nemo import lightning as nl
+from nemo.collections import llm
+from nemo.collections.llm.api import train
+from nemo.collections.llm.gpt.data import PreTrainingDataModule
+from nemo.collections.nlp.modules.common.tokenizer_utils import get_nmt_tokenizer
+from nemo.lightning import NeMoLogger
+from nemo.lightning.pytorch.callbacks import ModelCheckpoint
+
```
Comment (ericharper) [inline examples/llm/megatron_gpt_pretraining.py:12]: Can we add a comment here explaining the purpose of this script?

Comment (ashors1) [reply examples/llm/megatron_gpt_pretraining.py:15]: Done

**Review by ericharper — APPROVED** [review]
LGTM. Thanks!
